### PR TITLE
feat: add selector option to Makefile. Regenerate only namespace with changes

### PR DESCRIPTION
**Pull request by jalonsoa**
Add Makefile options to regenerete only namespaced with changes using helmfile selectors.

### CODE REVIEW

**Review by msvticket — CHANGES_REQUESTED** [review]
This change not a chore, but a new feature. So the commit message prefix should be changed from chore to feat.

You also need to resolve the conflict.

File: src/Makefile.mk [hunk 1]
```diff
@@ -149,6 +160,37 @@ fetch: init $(COPY_SOURCE) $(REPOSITORY_RESOLVE)
...
+
+# convert k8s Secrets => ExternalSecret resources using secret mapping + schemas
+# see: https://github.com/jenkins-x/jx-secret#mappings
+	jx secret convert --source-dir $(OUTPUT_DIR) -r $(VAULT_ROLE_EXTERNAL_SECRETS) -m ${VAULT_MOUNT_POINT_EXTERNAL_SECRETS}
+
+# replicate secrets to local staging/production namespaces
+	jx secret replicate --selector secret.jenkins-x.io/replica-source=true
+
+# populate secrets from filesystem definitions
+#	-VAULT_ADDR=$(VAULT_ADDR) VAULT_NAMESPACE=$(VAULT_NAMESPACE) jx secret populate --source filesystem --secret-namespace $(VAULT_NAMESPACE)
```
[the remark this answers is not in the document]
Comment (jalonsoa) [reply src/Makefile.mk:192]: From the helmfile help:

USAGE:
   helmfile [global options] command [command options] [arguments...]

The selector is a global option to helmfile, so it has to be put before the command. HELMFILE_TEMPLATE_FLAGS are options for template command, so we can't use the same variable for both cases.

I'm trying to solve this ... wait a moment .. please

[the remark this answers is not in the document]
Comment (jalonsoa) [reply src/Makefile.mk]: I create a unique option and then use this option to generate a new variable for clean step and other for fetch step.

HELMFILE_USE_SELECTORS ?= false

If is false:
   CLEAN_OUTPUT_DIR = $(OUTPUT_DIR)/*/*/
   HELMFILE_GLOBAL_FLAGS = ""

if true:
  CLEAN_OUTPUT_DIR = ""
  HELMFILE_GLOBAL_FLAGS = $(shell versionStream/src/get-selectors-and-clean.sh $(OUTPUT_DIR))

with this aproach, we have a unique clean step and a unique fetch .. it's ok ?

File: src/get-selectors-and-clean.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+OUTPUT_DIR=$1
+
+SELECTOR=""
+if [ -z "$(git log -m -1 --name-only --pretty=format: | grep -v "^helmfiles/")" ]; then
```
[the remark this answers is not in the document]
Comment (jalonsoa) [reply src/get-selectors-and-clean.sh]: done

File: src/Makefile.mk [hunk 3]
```diff
@@ -78,9 +78,16 @@ SOURCE_DIR ?= /workspace/source
 # HELMFILE_TEMPLATE_FLAGS ?= --debug
 HELMFILE_TEMPLATE_FLAGS ?=
 
+# this option configure use of selector to regenerate only namespaces with changes
+# by default, regenerate all. To regenerate only namespace try
+# HELMFILE_USE_SELECTORS = true
+HELMFILE_USE_SELECTORS ?= false
+CLEAN_OUTPUT_DIR ?= $(if $(findstring true,$(HELMFILE_USE_SELECTORS)),,$(OUTPUT_DIR)/*/*/)
+HELMFILE_GLOBAL_FLAGS ?= $(if $(findstring true,$(HELMFILE_USE_SELECTORS)),$(shell versionStream/src/get-selectors-and-clean.sh $(OUTPUT_DIR)),)
```
Comment (msvticket) [inline src/Makefile.mk:86]: With `?=` you make it possible to override this. It is in principle reasonable that a user would like to provide their own global flags for helmfile and I don't think that should break the use of selectors. I think you can just replace `?=` with `+=` to make this work.

**Review by msvticket — CHANGES_REQUESTED** [review]
Much better! This is easier to use as well. But I'd like one more usability fix.

File: src/Makefile.mk (hunk 3 above)
Comment (jalonsoa) [reply src/Makefile.mk]: of course !

**Review by msvticket — APPROVED** [review]

**Pull request merged**